dask/core.py

```diff
@@ -41,6 +41,122 @@ def istask(x):
+    # These nodes are guaranteed to not be part of a cycle.
+    completed = builtins.set()
+
+    # All nodes that have been visited in the current traversal.  Because
+    # we are doing depth-first search, going "deeper" should never result
+    # in visiting a node that has already been seen.  The `seen` and
+    # `completed` sets are mutually exclusive; it is okay to visit a node
+    # that has already been added to `completed`.
+    seen = builtins.set()
+
+    for key in keys:
+        if key in completed:
+            continue
+        nodes = [key]
+        while nodes:
+            # Keep current node on the stack until all descendants are visited
+            cur = nodes[-1]
+            if cur in completed:
+                # Already fully traversed descendants of cur
+                nodes.pop()
+                continue
+            seen.add(cur)
+
+            # Add direct descendants of cur to nodes stack
+            next_nodes = []
+            for nxt in get_dependencies(d, cur):
+                if nxt not in completed:
+                    if nxt in seen:
+                        # Cycle detected!
+                        return False
+                    next_nodes.append(nxt)
+
+            if next_nodes:
+                nodes.extend(next_nodes)
+            else:
+                # cur has no more descendants to explore, so we're done with it
+                completed.add(cur)
+                seen.remove(cur)
+                nodes.pop()
+    return True
+
+
+def _get_task(d, task, maxdepth=1000):
+    # non-recursive.  DAG property is checked upon reaching maxdepth.
+    depth = 0
+    _iter = lambda *args: iter(args)
+
+    # We construct a nested heirarchy of tuples to mimic the execution stack
+    # of frames that Python would maintain for a recursive implementation.
+    # A frame is associated with a single task from a Dask.
+    # A frame tuple has four elements:
+    #    1) The function for the task.
+    #    2) The arguments for the task (typically keys in the Dask).
+    #       Arguments are stored in reverse order, and elements are popped
+    #       as they are evaluated.
+    #    3) The calculated results of the arguments from (2).
+    #    4) The previous frame.
+    frame = (task[0], list(task[:0:-1]), [], None)
+    while True:
+        func, args, results, prev = frame
+        if not args:
+            val = func(*results)
+            if prev is None:
+                return val
+            prev[2].append(val)
+            frame = prev
+            depth -= 1
+            continue
+
+        key = args.pop()
+        if isinstance(key, list):
+            # v = (get(d, k, concrete=False) for k in key)  # recursive
+            # Fake being lazy
+            frame = (_iter, key[::-1], [], frame)
+            depth += 1
+            if maxdepth and depth > maxdepth:
+                if not isdag(d, list(task[1:])):
+                    raise ValueError('Cycle detected in dask!')
```

Comment:
Let's change this exception to either a dask-defined exception or a `RuntimeError`, which is what one would get with a recursive method.